### PR TITLE
[SVE] Check for SVE target in VectorizeLoop

### PR DESCRIPTION
**Pull request by ekalda**
Check that we are compiling for an SVE enabled target when the extent of a loop marked for vectorizing is a vscale dependent expression.

The extent of a loop should be either a positive integer or an vscale dependent expression, in the latter case we'd expect the target to have `has_sve` feature.

### CODE REVIEW

File: src/tir/transforms/vectorize_loop.cc [hunk 1]
```diff
@@ -725,17 +728,33 @@ class Vectorizer : public StmtMutator, public ExprFunctor<PrimExpr(const PrimExp
...
+      }
       ICHECK(is_zero(op->min));
       return Vectorizer(op->loop_var, op->extent)(op->body);
     } else {
       return StmtMutator::VisitStmt_(op);
     }
   }
-};
 
-Stmt VectorizeLoop(Stmt stmt) { return LoopVectorizer()(std::move(stmt)); }
```
Comment (ekalda) [inline src/tir/transforms/vectorize_loop.cc:738]: Is this function necessary?

**Comment by ekalda** [comment]
@Lunderberg @lhutton1 

This is a follow up from https://github.com/apache/tvm/pull/16782, it makes sure that scalable vectors are created only for SVE target. In this implementation we check for target in function attributes, so we'd rely on `BindTarget` having run before. Alternatively we could access the "global" target via `Target::Current()`, which would be simpler, but would make it harder to mix SVE and non-SVE targets. I'm happy to go with either option :)

**Comment by ekalda** [comment]
I moved the implementation to use the current target since changing the graph executor's pipeline is breaking lots of CUDA tests.

File: tests/python/tir-transform/test_tir_transform_vectorize.py [hunk 2]
```diff
@@ -99,26 +105,29 @@ def main(A: T.Buffer((25,), "float32")):
...
+        with tvm.target.Target(sve_target):
+            tvm.tir.transform.VectorizeLoop()(Module)
 
 
 def test_vectorize_vector_scalable_error4():
     @I.ir_module
     class Module:
         @T.prim_func(private=True)
         def main(A: T.Buffer((25,), "float32")):
+            T.func_attr({"target": sve_target})
```
Comment (lhutton1) [inline tests/python/tir-transform/test_tir_transform_vectorize.py:117]: Is this still needed?

File: src/tir/transforms/vectorize_loop.cc [hunk 3]
```diff
@@ -727,6 +730,17 @@ class LoopVectorizer : public StmtMutator {
  public:
   Stmt VisitStmt_(const ForNode* op) final {
     if (op->kind == ForKind::kVectorized) {
+      auto* extent_as_int = op->extent.as<IntImmNode>();
+      if (!extent_as_int || extent_as_int->value < 1) {
+        Target current_target = Target::Current();
+        bool has_sve{false};
+        if (current_target.defined()) {
+          has_sve = current_target->GetFeature<Bool>("has_sve").value_or(Bool(false));
```
[the remark this answers is not in the document]
Comment (ekalda) [reply src/tir/transforms/vectorize_loop.cc:738]: Good idea, I'll create a utility for it, it probably won't be the last time we would need to do that check :) There is another new utility function in @Anndrey24's [patch that is in progress ](https://github.com/apache/tvm/pull/16899/files#diff-a31e6da5360b6615078023bb62667d79ff77fad68c29dda820a9267a918d9c01L44) for checking whether an expression contains `vscale`, so once this goes in, we should unify the uses for that case as well.

File: tests/python/tir-transform/test_tir_transform_vectorize.py (hunk 2 above)
Comment (ekalda) [reply tests/python/tir-transform/test_tir_transform_vectorize.py:117]: Oops no, will remove.

**Review by lhutton1 — APPROVED** [review]
Thanks @ekalda, LGTM!

**Comment by lhutton1** [comment]
Thanks @ekalda!

**Pull request merged**